### PR TITLE
Improve listener handling and fix tests

### PR DESCRIPTION
**Pull request by mpenick**
While improving the close handling of the listener in `Proxy` I noticed
that this could be generalized to handle multiple listeners (like
`http.Server`). This could be quite useful in the future for supporting
different multi-proxy features in a more user-friendly and resource
efficient way.  So `Proxy`, the type, now allows multiple listeners in a
single proxy. This hasn't been exposed in an external feature yet.

This also improves clean up for servers, and their listeners, used
during testing and now tests generate unique, unbound ports for each
test to avoid sporadic errors I was seeing in CI.

### CODE REVIEW

**Comment by mpenick** [comment]
Most of the changes are in tests, and the most critical changes are in `proxy/proxy.go` and `proxy/run.go`.

**Review by joao-r-reis — APPROVED** [review]
Interesting work to make the listener part of the proxy more pluggable but do you already have a use case in mind for this? What was the motivation behind this work?

File: proxy/run.go [hunk 1]
```diff
@@ -263,21 +263,25 @@ func listenAndServe(p *Proxy, mux *http.ServeMux, ctx context.Context, logger *z
 
 	// Listen is called first to set up the listening server connection and establish initial client connections to the
 	// backend cluster so that when the readiness check is hit the proxy is actually ready.
-	err = p.Listen(config.Bind)
+
+	err = p.Connect()
```
[the remark this answers is not in the document]
Comment (mpenick) [reply proxy/run.go:267]: This completely decouples listening from serving and the backend cluster. The cool part about doing that is you can have multiple listeners or listeners on "unix" sockets if you wanted...

**Comment by mpenick** [comment]
> Interesting work to make the listener part of the proxy more pluggable but do you already have a use case in mind for this? What was the motivation behind this work?

Most of that started with making listener cleanup more robust, but then it's easy to generalize to multiple listeners, which could be used for bind multiple nodes in a single proxy (useful for multi-region in some setups).

**Pull request merged**